### PR TITLE
virtual groups

### PR DESCRIPTION
**Pull request by ArtificialOwl**
### Refactoring of the app.

The app will now returns result on strict comparison of entities with access to resources. It is working based on the type of those entities to avoid strange behavior on cross shares:

- If only single-user entities (ie. users) have access to the resource, it will create a virtual group of those users including shares recipients and share initiator/resources owner
- If a multi-user entity (ie. groups, circles) have access to the resource, it will ignore single entity and only compare the multi-user entity that have access to the resource (shares recipients)

As discussed previously, there is actually no efficient way to mix and compare single entities and multi-user entities.

### tasks:

- [x] full refactoring of the app,
- [x] files,
- [x] deck,
- [x] talk,
- [x] calendar,
- [x] caching,
- [x] flush cache on new shares
- [ ] complete testing of app in huge env.

There is actually no efficient way to remove the right cached elements on share creation/deletion, so any changes in share will flush all cache from the app.

### CODE REVIEW

File: lib/RelatedResourceProviders/FilesRelatedResourceProvider.php [hunk 1]
```diff
@@ -233,4 +200,36 @@ private function getItemIdsFromParentPath(int $itemId): array {
...
+			$sharedWith = $this->miscService->convertShareRecipient(
+				$share->getShareType(),
+				$share->getSharedWith()
+			);
+
+			if ($share->getShareType() === IShare::TYPE_USER) {
+				$shareCreator = $this->miscService->convertShareRecipient(
+					IShare::TYPE_USER,
+					$share->getShareCreator()
+				);
```
Comment (come-nc) [inline lib/RelatedResourceProviders/FilesRelatedResourceProvider.php:220]: Why is creator added to virtual group for files but not deck or calendar?

File: lib/RelatedResourceProviders/FilesRelatedResourceProvider.php [hunk 2]
```diff
@@ -233,4 +200,36 @@ private function getItemIdsFromParentPath(int $itemId): array {
...
+	 */
+	private function processShareRecipient(RelatedResource $related, FilesShare $share) {
+		try {
+			$sharedWith = $this->miscService->convertShareRecipient(
+				$share->getShareType(),
+				$share->getSharedWith()
+			);
+
+			if ($share->getShareType() === IShare::TYPE_USER) {
+				$shareCreator = $this->miscService->convertShareRecipient(
```
Comment (come-nc) [inline lib/RelatedResourceProviders/FilesRelatedResourceProvider.php:217]: `miscService` ?

File: lib/Service/RelatedService.php [hunk 3]
```diff
@@ -120,104 +127,114 @@ public function getRelatedToItem(string $providerId, string $itemId, int $chunk
...
 		try {
-			$shares = $this->getCachedSharesRecipients($providerId, $itemId);
-
-			return $shares;
+			return $this->getCachedRelatedFromItem($providerId, $itemId);
 		} catch (CacheNotFoundException $e) {
 		}
 
+		$this->getRelatedResourceProvider($providerId);
 		$result = $this->getRelatedResourceProvider($providerId)
```
Comment (come-nc) [inline lib/Service/RelatedService.php:228]:
```suggestion
		$result = $this->getRelatedResourceProvider($providerId)
```
Unless I misunderstood the first call is useless?

File: lib/RelatedResourceProviders/FilesRelatedResourceProvider.php (hunk 1 above)
Comment (ArtificialOwl) [reply lib/RelatedResourceProviders/FilesRelatedResourceProvider.php]: The owner of the share is not defined the same way for deck/calendar, so I am using the owner of the entry: https://github.com/nextcloud/related_resources/blob/fc1bf6216003cc2374d8d5d95a77153336eb7510/lib/RelatedResourceProviders/DeckRelatedResourceProvider.php#L99

File: lib/RelatedResourceProviders/FilesRelatedResourceProvider.php (hunk 2 above)
Comment (ArtificialOwl) [reply lib/RelatedResourceProviders/FilesRelatedResourceProvider.php]: deprecated

File: lib/IRelatedResourceProvider.php [hunk 4]
```diff
@@ -44,20 +44,18 @@ public function getProviderId(): string;
...
-	public function getSharesRecipients(string $itemId): array;
+	public function getRelatedFromItem(string $itemId): ?IRelatedResource;
 
 	/**
-	 * returns all shares done to multiple entities
-	 *
 	 * @param FederatedUser $entity
 	 *
-	 * @return IRelatedResource[]
+	 * @return string[]
```
[the remark this answers is not in the document]
Comment (ArtificialOwl) [reply lib/IRelatedResourceProvider.php:58]: commenting

**Pull request merged**